test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs

```diff
@@ -1118,13 +1122,85 @@ public void ProviderCacheTest_EnsureNoException_MultipleThreads()
+                new Claim(ClaimTypes.AuthenticationInstant, Default.AuthenticationInstant, ClaimValueTypes.DateTime, Default.Issuer)
+            };
+
+            var header = new JwtHeader(Default.SymmetricEncryptingCredentials);
+            JwtPayload payload = new JwtPayload("IssuerName", "Audience", testClaims, DateTime.Now.AddHours(-1), DateTime.Now.AddHours(1), DateTime.Now.AddHours(-1));
+            var token = new JwtSecurityToken(header, payload);
+
+            string certHash = "Test Cert Hash";
+            token.Header[JwtHeaderParameterNames.X5t] = certHash;
+            token.Header[JwtHeaderParameterNames.Kid] = certHash;
+
+            var handler = new JwtSecurityTokenHandler();
+            _ = handler.WriteToken(token);
+
+            // when JwtHeader is created with SymmetricEncryptingCredentials, the provider will not be added to cache (an error in logic???)
+            var cache = CryptoProviderFactory.Default.CryptoProviderCache as InMemoryCryptoProviderCache;
+            Assert.True(cache.TaskCount == 0, $"ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SymmetricEncryptingCredentials: unexpected task count: {cache.TaskCount}, expected: 0");
+        }
+
+        /// <summary>
+        /// Test to ensure no hanging task at the end when calling the JwtSecurityTokenHandler.WriteToken() method.
+        /// The JwtHeder is created with SigningCredentials.
+        /// </summary>
+        [Fact]
+        public void ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SigningCredentials()
+        {
+            var testClaims = new List<Claim>
+            {
+                new Claim(ClaimTypes.AuthenticationMethod, Default.AuthenticationMethod, ClaimValueTypes.String, Default.Issuer),
+                new Claim(ClaimTypes.AuthenticationInstant, Default.AuthenticationInstant, ClaimValueTypes.DateTime, Default.Issuer)
+            };
+
+            var signingCredentials = new SigningCredentials(new X509SecurityKey(KeyingMaterial.DefaultCert_2048), SecurityAlgorithms.RsaSha256);
+
+            var token = new JwtSecurityToken(
+                issuer: "IssuerName",
+                audience: "Audience",
+                claims: testClaims,
+                notBefore: DateTime.Now.AddHours(-1),
+                expires: DateTime.Now.AddHours(1),
+                signingCredentials: signingCredentials);
+
+            string certHash = "Test Cert Hash";
+
+            token.Header[JwtHeaderParameterNames.X5t] = certHash;
+            token.Header[JwtHeaderParameterNames.Kid] = certHash;
+
+            var handler = new JwtSecurityTokenHandler();
+            _ = handler.WriteToken(token);
+
+            int waitTimeoutInSeconds = 180;
+
+            var cache = CryptoProviderFactory.Default.CryptoProviderCache as InMemoryCryptoProviderCache;
+
+            Thread.Sleep(1000); // wait one second to allow the task to be successfully started
```

Comment:
>Thread.Sleep(1000);  [](http://example.com/codeflow?start=12&length=20)

I'm not good with thread.sleep in tests. 
Should be more deterministic. (an internal flag on the cache would be better imo)

Reply:
Yeah, sounds good.